### PR TITLE
[Alerting]: Add alerting endpoint for Query Evaluation

### PR DESCRIPTION
**Pull request by papagian**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes grafana/alerting-squad#94

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/services/ngalert/api/test-data/test.http [hunk 1]
```diff
@@ -25,6 +25,52 @@ content-type: application/json
...
+  },
+  {
+    "refId": "B",
+    "queryType": "",
+    "relativeTimeRange": {
+      "from": 18000,
+      "to": 10800
+    },
+    "model": {
+      "datasource": "__expr__",
```
Comment (kylebrandt) [inline pkg/services/ngalert/api/test-data/test.http:60]: should drop this and just be uid

File: pkg/services/ngalert/api/test-data/test.http [hunk 2]
```diff
@@ -25,6 +25,52 @@ content-type: application/json
...
+      "datasourceUid": "__expr__",
```
Comment (kylebrandt) [inline pkg/services/ngalert/api/test-data/test.http:61]: uid should be "-100"

**Comment by kylebrandt** [comment]
Seems like from the time returned that "now" isn't working, but not sure why.

Example:

```
POST http://admin:admin@localhost:3000/api/v1/eval
content-type: application/json

{
"data": [
  {
    "refId": "A",
    "queryType": "",
    "relativeTimeRange": {
      "from": 500,
      "to": 0
    },
    "model": {
      "datasourceUid": "000000004",
      "intervalMs": 1000,
      "maxDataPoints": 100,
      "orgId": 0,
      "refId": "A",
      "scenarioId": "csv_metric_values",
      "stringInput": "1,20,90,30,5,0"
    }
  },
  {
    "refId": "B",
    "queryType": "",

    "model": {
      "datasourceUid": "-100",
      "expression": "$A",
      "intervalMs": 2000,
      "maxDataPoints": 200,
      "orgId": 0,
      "reducer": "mean",
      "refId": "B",
      "type": "reduce"
    }
  }
],
"now": "2021-04-11T14:38:14Z"
}
```

Timestamps in the response for A don't seem to line up:

```
          "data": {
            "values": [
              [
                1618936212000,
                1618936312000,
                1618936412000,
                1618936512000,
                1618936612000,
                1618936712000
```

File: pkg/services/ngalert/api/api_testing.go [hunk 3]
```diff
@@ -77,3 +78,18 @@ func (srv TestingApiSrv) RouteTestRuleConfig(c *models.ReqContext, body apimodel
...
 }
+
+func (srv TestingApiSrv) RouteEvalQueries(c *models.ReqContext, cmd apimodels.EvalQueriesPayload) response.Response {
+	now := cmd.Now
+	if now.IsZero() {
+		now = timeNow()
+	}
+
+	evaluator := eval.Evaluator{Cfg: srv.Cfg}
+	evalResults, err := evaluator.QueriesAndExpressionsEval(c.SignedInUser.OrgId, cmd.Data, timeNow(), srv.DataService)
```
Comment (kylebrandt) [inline pkg/services/ngalert/api/api_testing.go:89]: I think maybe this should be `now` and not `timeNow()`

Comment (kylebrandt) [reply pkg/services/ngalert/api/api_testing.go:89]: (This fixes what I noted before based on my manual tests).

Comment (papagian) [reply pkg/services/ngalert/api/api_testing.go:89]: thanks for finding it!

**Review by kylebrandt — APPROVED** [review]
Still http file to update, but other than that looks good to me. Thanks for doing this!

File: pkg/services/ngalert/api/api_testing.go [hunk 4]
```diff
@@ -77,3 +78,18 @@ func (srv TestingApiSrv) RouteTestRuleConfig(c *models.ReqContext, body apimodel
 		nil,
 	)
 }
+
+func (srv TestingApiSrv) RouteEvalQueries(c *models.ReqContext, cmd apimodels.EvalQueriesPayload) response.Response {
```
Comment (marefr) [inline pkg/services/ngalert/api/api_testing.go:82]: Is there proper permission checks in place to make sure someone cannot query a data source that they don't have permission to?

**Review by marefr — COMMENTED** [review]
Added a comment. Other than that LGTM

File: pkg/services/ngalert/api/api_testing.go (hunk 4 above)
Comment (kylebrandt) [reply pkg/services/ngalert/api/api_testing.go:82]: Good catch I think not. Currently SSE is using the Bus to lookup data sources.

I think I need to update SSE to take a signed in user and data source cache to ensure it is always checked close to the query.

Comment (papagian) [reply pkg/services/ngalert/api/api_testing.go:82]: good point! I think that the way datasource permissions are applied (given it's an enterprise feature) is that they override the datasource cache. In the similar endpoint for evaluating a condition I used to validate if all the queries referenced an existing datasource therefore I had [to query](https://github.com/grafana/grafana/blob/0dcd0281abbbb962697cca9ad30dd5ac43e1f859/pkg/services/ngalert/api/util.go#L185) the cache. I will add a similar check for this endpoint.

Comment (marefr) [reply pkg/services/ngalert/api/api_testing.go:82]: Sounds good :+1:

**Comment by papagian** [comment]
I have deleted the additional tests and I will re-introduced them to a new PR because resolving the conflicts and fixing the timeouts is hard at once.

**Pull request merged**